scripts/hotWalletCli/client.ts

```diff
@@ -0,0 +1,283 @@
+
+        const block = await this.rpc.getBlock({ blockNumber })
+
+        const timeDifferenceSeconds = targetTimestamp - block.timestamp
+
+        // Block is within 1 block before the target timestamp
+        if (timeDifferenceSeconds >= 0n && timeDifferenceSeconds <= averageBlockTimeSeconds) break
+
+        const blocksToMove = BigInt(Math.ceil(Math.abs(Number(timeDifferenceSeconds)) / averageBlockTimeSeconds))
+
+        if (block.timestamp > targetTimestamp) {
+          blockNumber -= blocksToMove
+        } else {
+          blockNumber += blocksToMove
+        }
+      }
+
+      // In case of multiple batched blocks for a target timestamp, find the earliest or latest block based on blockMode
+      while (true) {
+        if (blockNumber <= 0n) return 0n
+        if (blockNumber >= latestBlock.number) return latestBlock.number
+
+        const nextBlockNumber = blockMode === 'earliest' ? blockNumber - 1n : blockNumber + 1n
+
+        const block = await this.rpc.getBlock({ blockNumber: nextBlockNumber })
+
+        if (block.timestamp !== targetTimestamp) break
+
+        blockNumber = nextBlockNumber
+      }
+
+      return blockNumber
+    } catch (err) {
+      if (err instanceof Error) {
+        const text = `Failed to get block for timestamp: ${targetTimestamp}: ${err.message}, exiting.`
+        spinner ? spinner.fail(text) : error(text)
+      } else {
+        const text = `Failed to get block for timestamp: ${targetTimestamp}, exiting.`
+        spinner ? spinner.fail(text) : error(text)
+      }
+
+      process.exit(1)
+    }
+  }
+
+  async getRevenue(startTimestamp: number, endTimestamp: number): Promise<Revenue> {
+    try {
+      const { data } = await axios.get<Revenue>(
+        `https://api.thorchain.shapeshift.com/api/v1/affiliate/revenue?start=${startTimestamp}&end=${endTimestamp}`,
+      )
+      return data
+    } catch (err) {
+      if (isAxiosError(err)) {
+        error(
+          `Failed to get revenue for period (start: ${startTimestamp} - end: ${endTimestamp}): ${err.message}, exiting.`,
+        )
+      } else {
+        error(`Failed to get revenue for period (start: ${startTimestamp} - end: ${endTimestamp}), exiting.`)
+      }
+
+      process.exit(1)
+    }
+  }
+
+  private async getClosingStateByStakingAddress(
+    addresses: Address[],
+    startBlock: bigint,
+    endBlock: bigint,
+  ): Promise<Record<string, ClosingState>> {
+    const contract = getContract({
+      address: ARBITRUM_RFOX_PROXY_CONTRACT_ADDRESS,
+      abi: stakingV1Abi,
+      client: { public: this.rpc },
+    })
+
+    const prevEpochEndBlock = startBlock - 1n
+
+    const closingStateByStakingAddress: Record<string, ClosingState> = {}
+    for await (const address of addresses) {
+      const [stakingBalance, _unstakingBalance, _earnedRewards, _rewardPerTokenStored, runeAddress] =
+        await contract.read.stakingInfo([getAddress(address)], { blockNumber: endBlock })
+
+      if (stakingBalance <= 0n) continue
```

Comment:
Does this handle people who have withdrawn before EOM but are still due rewards properly? 

It seems like it would skip them.

Reply:
Good catch, actioned here: fb4227d50539082960b42f2465e5b2c4c2a36fa0 and added the condition to omit addresses with no reward units earned for the period